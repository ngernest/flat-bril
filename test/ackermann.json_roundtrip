{
  "functions": [
    {
      "args": [
        {
          "name": "m",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "m",
            "zero"
          ],
          "dest": "cond_m",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "cond_m"
          ],
          "funcs": [],
          "labels": [
            "m_zero",
            "m_nonzero"
          ],
          "op": "br"
        },
        {
          "label": "m_zero"
        },
        {
          "args": [
            "n",
            "one"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "m_nonzero"
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "cond_n",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "cond_n"
          ],
          "funcs": [],
          "labels": [
            "n_zero",
            "n_nonzero"
          ],
          "op": "br"
        },
        {
          "label": "n_zero"
        },
        {
          "args": [
            "m",
            "one"
          ],
          "dest": "m1",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "m1",
            "one"
          ],
          "dest": "tmp",
          "funcs": [
            "ack"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "n_nonzero"
        },
        {
          "args": [
            "m",
            "one"
          ],
          "dest": "m1",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "n",
            "one"
          ],
          "dest": "n1",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "m",
            "n1"
          ],
          "dest": "t1",
          "funcs": [
            "ack"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "m1",
            "t1"
          ],
          "dest": "t2",
          "funcs": [
            "ack"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "t2"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "ack",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "m",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "m",
            "n"
          ],
          "dest": "tmp",
          "funcs": [
            "ack"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
