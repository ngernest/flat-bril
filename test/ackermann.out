@ack
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: eq   		dest: cond_m		type: bool		args: ["m", "zero"]	
	op: br   		args: ["cond_m"]		labels: ["m_zero", "m_nonzero"]
.m_zero
	op: add  		dest: tmp		type: int		args: ["n", "one"]	
	op: ret  		args: ["tmp"]	
.m_nonzero
	op: eq   		dest: cond_n		type: bool		args: ["n", "zero"]	
	op: br   		args: ["cond_n"]		labels: ["n_zero", "n_nonzero"]
.n_zero
	op: sub  		dest: m1		type: int		args: ["m", "one"]	
	op: call 		dest: tmp		type: int		args: ["m1", "one"]		funcs: ["ack"]
	op: ret  		args: ["tmp"]	
.n_nonzero
	op: sub  		dest: m1		type: int		args: ["m", "one"]	
	op: sub  		dest: n1		type: int		args: ["n", "one"]	
	op: call 		dest: t1		type: int		args: ["m", "n1"]		funcs: ["ack"]
	op: call 		dest: t2		type: int		args: ["m1", "t1"]		funcs: ["ack"]
	op: ret  		args: ["t2"]	
@main
	op: call 		dest: tmp		type: int		args: ["m", "n"]		funcs: ["ack"]
	op: print		args: ["tmp"]	
